File: golang/codec/encoder.go
package codec

import (
	"fmt"
	"io"
	"sync"
)

type Encoder struct {
	w io.Writer
	sync.Mutex
}

func NewEncoder(w io.Writer) *Encoder {
	return &Encoder{w: w}
}

func (enc *Encoder) Encode(msg interface{}) error {
	enc.Lock()
	defer enc.Unlock()

	if DebugMessages != nil {
		fmt.Fprintln(DebugMessages, "<<ENC", msg)
	}

	b, err := Marshal(msg)
	if err != nil {
		return err
	}

	if DebugBytes != nil {
		fmt.Fprintln(DebugBytes, "<<ENC", b)
	}

	_, err = enc.w.Write(b)
	return err
}

type Marshaler interface {
	MarshalMux() ([]byte, error)
}

func Marshal(v interface{}) ([]byte, error) {
	m, ok := v.(Marshaler)
	if !ok {
		return []byte{}, fmt.Errorf("qmux: unable to marshal type")
	}
	return m.MarshalMux()
}

File: csharp/qmux/codec/encoder.cs
namespace qmux.codec;

using System.Text;
using System.Threading;
using gostdlib.errors;
using gostdlib.io;

public class Encoder
{
    private Mutex mutex = new Mutex();
    private io.IWriter w;

    public Encoder(io.IWriter w)
    {
        this.w = w;
    }

    public errors.Error? Encode(object msg)
    {
        this.mutex.WaitOne();
        if (Debug.Messages != null)
        {
            var debugBytes = Encoding.UTF8.GetBytes("<<ENC" + msg.ToString());
            Debug.Messages.Write(debugBytes);
        }

        var (b, err) = Marshaler.Marshal(msg);
        if (err != null)
        {
            return err;
        }

        var _ = this.w.Write(b);
        if (Debug.Bytes != null)
        {
            var debugBytes = Encoding.UTF8.GetBytes("<<ENC" + b);
            Debug.Bytes.Write(debugBytes);
        }

        this.mutex.ReleaseMutex();
        return null;
    }
}